apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -606,6 +609,87 @@ mod prepare_block {
             // assert_eq!(rsp.tx_records, vec![digest]);
         }
 
+        /// Test if vote extension validation and inclusion in a block
+        /// behaves as expected, considering <= 2/3 voting power.
+        #[test]
+        // TODO: make sure this test panics
+        //#[should_panic(expected = "entered unreachable code")]
+        fn test_prepare_proposal_vext_insufficient_voting_power() {
+            const LAST_HEIGHT: BlockHeight = BlockHeight(3);
+
+            // starting the shell like this will contain insufficient voting
+            // power
+            let (mut shell, _, _) = test_utils::setup();
+
+            // artificially change the block height
+            shell.storage.last_height = LAST_HEIGHT;
+
+            // artificially change the voting power of the default validator to
+            // 0
+            //
+            // TODO: this is not working
+            let events_epoch = shell
+                .storage
+                .block
+                .pred_epochs
+                .get_epoch(LAST_HEIGHT)
+                .expect("Test failed");
+            let validator_set = {
+                let params = shell.storage.read_pos_params();
+                let mut epochs = shell.storage.read_validator_set();
+                let mut data =
+                    epochs.get(events_epoch).cloned().expect("Test failed");
+
+                data.active = data
+                    .active
+                    .iter()
+                    .cloned()
+                    .map(|v| WeightedValidator {
+                        voting_power: VotingPower::from(0u64),
+                        ..v
+                    })
+                    .collect();
+
+                epochs.set(data, events_epoch, &params);
+                epochs
+            };
+            shell.storage.write_validator_set(&validator_set);
```

Comment:
I'm not sure why this is not working as expected (the one default validator would get a voting power of 0). This would provoke the `prepare_proposal` call to reach `unreachable!()` code, as expected (since we would get <= 2/3 voting power).

Any ideas @james-chf @batconjurer ??

Reply:
```
let mut req = FinalizeBlock::default();
            req.header.time = namada::types::time::DateTimeUtc::now();
            shell.storage.last_height = BlockHeight(11);
            shell.finalize_block(req).expect("Test failed");
            shell.commit();
            ``

Reply:
validator set changes don't take effect unless the epoch changes. You've only queued up the change